packages/@lwc/engine-core/src/libs/mutation-tracker/index.ts

```diff
@@ -102,14 +102,17 @@ export class ReactiveObserver {
         if (len > 0) {
             for (let i = 0; i < len; i++) {
                 const set = listeners[i];
-                if (set.length === 1) {
-                    // Perf optimization for the common case - the length is usually 1, so avoid the indexOf+splice.
+                const setLength = set.length;
+                if (setLength === 1) {
+                    // Perf optimization for the common case - the length is usually 1, so avoid the indexOf+push.
                     // If the length is 1, we can also be sure that `this` is the first item in the array.
                     set.length = 0;
                 } else {
-                    // Slow case
-                    const pos = ArrayIndexOf.call(set, this);
-                    ArraySplice.call(set, pos, 1);
+                    // Slow case - swap with the last item and then remove.
+                    // (Avoiding splice here is a perf optimization, and the order doesn't matter.)
+                    const index = ArrayIndexOf.call(set, this);
+                    set[index] = set[setLength - 1];
+                    ArrayPop.call(set);
```

Comment:
Is there a performance difference between `set.pop()` and `set.length = setLength - 1`? We use the former here, but the latter in the `if` block.

Reply:
That's a good catch. We can actually refactor the code to simplify it, if we're always just `pop`ing or setting `set.length = length - 1` in every case.

76f07c42f

I did a quick perf test, and there's basically no difference between `set.length = length - 1`:

![Screenshot 2024-04-10 at 2 00 59 PM](https://github.com/salesforce/lwc/assets/283842/502e19b4-51b2-4975-9d2d-84bdb69dc415)

... and `pop()`:

![Screenshot 2024-04-10 at 2 06 21 PM](https://github.com/salesforce/lwc/assets/283842/55e84830-af20-4077-aa2f-0da07ea30cde)

Although maybe `pop`()` is slightly faster? Either way I would prefer `pop()` since it's more idiomatic.

I'm also planning to run this through our full benchmark suite tonight, just in case there is some weird edge case I missed.

Reply:
Huh, so it looks like `pop()` is faster than setting `length`. In a minimal comparison, the benchmark reports that `pop()` is 1-4% faster:

![Screenshot 2024-04-10 at 3 14 06 PM](https://github.com/salesforce/lwc/assets/283842/567c2bb6-1c7b-408f-9be3-a21b0c3ff54e)